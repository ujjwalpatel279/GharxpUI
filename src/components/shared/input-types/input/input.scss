// ========================================================================
// COMPONENTS/SHARED/FORM-INPUT
// ========================================================================

@import '~sassHelpers';

.c-input {
  width: 100%;
  height: spacer(10);
  padding-left: spacer(4);
  padding-right: spacer(4);
  border: 1px solid color(grey3);
  border-radius: $border-radius;
  background-color: white;

  &.isRequired {
    border-left: 5px solid color(navy1);
  }

  &.isInvalid {
    background-color: color(errorLight);
    border-color: color(errorDark);
  }

  &.hasValue {
    background-color: color-brand(primary-light-2);
  }

  &:hover,
  &:focus {
    border-color: var(--primary);
  }
}
