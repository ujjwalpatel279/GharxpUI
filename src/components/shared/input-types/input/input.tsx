import React, { ChangeEvent, forwardRef, KeyboardEvent, useEffect } from 'react';
import { useForm } from 'react-hook-form';

import './input.scss';
interface InputProps {
  id: string;
  type?: 'text' | 'number' | 'email' | 'tel' | 'search';
  placeHolder?: string;
  required?: boolean;
  disabled?: boolean;
  error?: boolean;
  childClasses?: string;
  onKeyUp?: (e: KeyboardEvent<HTMLInputElement>) => void;
  onChange?: (e: ChangeEvent<HTMLInputElement>) => void;
  setValue?: ReturnType<typeof useForm>['setValue'];
  value?: string | number | boolean;
  name?: string;
}

type Ref = HTMLInputElement;

export const Input = forwardRef<Ref, InputProps>(
  (
    {
      id,
      type = 'text',
      placeHolder,
      onChange,
      onKeyUp,
      setValue,
      disabled,
      required,
      error,
      childClasses,
      value,
      name,
    },
    ref,
  ) => {
    let inputClasses = 'c-input';

    useEffect(() => {
      if (value) {
        setValue(id, value);
      }
    }, [value]);

    if (disabled) {
      inputClasses += ' isDisabled';
    }

    if (error) {
      inputClasses += ' isInvalid';
    }

    if (required) {
      inputClasses += ' isRequired';
    }

    if (value) {
      inputClasses += ' hasValue';
    }

    if (childClasses) {
      inputClasses += ` ${childClasses}`;
    }

    return (
      <input
        type={type}
        id={id}
        name={name ?? id}
        placeholder={placeHolder}
        className={inputClasses}
        ref={ref}
        aria-invalid={error ? 'true' : 'false'}
        required={required}
        disabled={disabled}
        onKeyUp={onKeyUp}
        onChange={onChange}
      />
    );
  },
);

Input.displayName = 'Input';
