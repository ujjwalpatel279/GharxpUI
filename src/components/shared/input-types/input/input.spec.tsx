import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';

import { Input } from './input';

describe('Input Component', () => {
  afterEach(cleanup);

  it('renders the form input component', () => {
    const { container } = render(<Input id="firstName" required={false} ref={() => null} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('aria-invalid is false', () => {
    const { queryByPlaceholderText } = render(
      <Input id="firstName" error={false} placeHolder="testInput" required={false} ref={() => null} />,
    );

    expect(queryByPlaceholderText('testInput').getAttribute('aria-invalid')).toBe('false');
  });

  it('aria-invalid is true', () => {
    const { queryByPlaceholderText } = render(
      <Input id="firstName" error={true} placeHolder="testInput" required={false} ref={() => null} />,
    );

    expect(queryByPlaceholderText('testInput').getAttribute('aria-invalid')).toBe('true');
  });

  it('input value is initially empty and has a value once a value is entered', () => {
    const { queryByPlaceholderText } = render(
      <Input id="firstName" error={true} placeHolder="testInput" required={false} ref={() => null} />,
    );

    const input = queryByPlaceholderText('testInput') as HTMLInputElement;
    expect(input.value).toBe('');

    fireEvent.change(input, { target: { value: 'Rafael' } });
    expect(input.value).toBe('Rafael');
  });
});
