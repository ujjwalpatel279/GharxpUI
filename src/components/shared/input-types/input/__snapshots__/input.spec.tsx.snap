// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component renders the form input component 1`] = `
<input
  aria-invalid="false"
  class="c-input"
  id="firstName"
  name="firstName"
  type="text"
/>
`;
